aquadoggo/src/db/stores/document.rs

```diff
@@ -0,0 +1,331 @@
+use p2panda_rs::document::{DocumentView, DocumentViewHash, DocumentViewId};
+use p2panda_rs::operation::OperationValue;
+use p2panda_rs::schema::SchemaId;
+use sqlx::{query, query_as};
+
+use crate::db::errors::DocumentViewStorageError;
+use crate::db::models::operation::OperationFieldRow;
+use crate::db::provider::SqlStorage;
+use crate::db::traits::{
+    AsStorageDocumentView, DocumentStore, DocumentViewFields, FieldIds, FieldName,
+};
+use crate::db::utils::parse_operation_fields;
+
+/// Aquadoggo struct which will implement AsStorageDocumentView trait.
+#[derive(Debug, Clone)]
+pub struct DocumentViewStorage {
+    document_view: DocumentView,
+    field_ids: FieldIds,
+    schema_id: SchemaId,
+}
+
+impl DocumentViewStorage {
+    pub fn new(document_view: &DocumentView, field_ids: &FieldIds, schema_id: &SchemaId) -> Self {
+        Self {
+            document_view: document_view.clone(),
+            field_ids: field_ids.clone(),
+            schema_id: schema_id.clone(),
+        }
+    }
+}
+
+impl AsStorageDocumentView for DocumentViewStorage {
+    type AsStorageDocumentViewError = DocumentViewStorageError;
+
+    fn id(&self) -> DocumentViewId {
+        self.document_view.id().clone()
+    }
+
+    fn iter(&self) -> Iter<FieldName, OperationValue> {
+        self.document_view.iter()
+    }
+
+    fn get(&self, key: &str) -> Option<&OperationValue> {
+        self.document_view.get(key)
+    }
+
+    fn schema_id(&self) -> SchemaId {
+        self.schema_id.clone()
+    }
+
+    fn field_ids(&self) -> FieldIds {
+        self.field_ids.clone()
+    }
+}
+
+#[async_trait]
+impl DocumentStore<DocumentViewStorage> for SqlStorage {
+    /// Insert a document_view into the db. Requires that all relevent operations are already in
+    /// the db as this method only creates relations between document view fields and their current
+    /// values (last updated operation value).
+    ///
+    /// QUESTION: Is this too implementation specific? It assumes quite a lot about the db
+    /// structure and others may wish to structure things differently.
+    async fn insert_document_view(
+        &self,
+        document_view_id: &DocumentViewId,
+        field_ids: &FieldIds,
+        schema_id: &SchemaId,
+    ) -> Result<bool, DocumentViewStorageError> {
+        // OBSERVATIONS:
+        // - we need to know which operation was LWW for each field.
+        // - this is different from knowing the document view id, which is
+        // just the tip(s) of the graph, and will likely not contain a value
+        // for every field.
+        // - we could record the operation id for each value when we build the
+        //   document
+        // - alternatively we could do some dynamic "reverse" graph traversal
+        // starting from the document view id. This would require
+        // implementing some new traversal logic (maybe it is already underway
+        // somewhere? I remember @cafca working on this a while ago).
```

Comment:
Yeah, I was _so_ sure that @cafca was working on this and even merged it, but now I can't find it anywhere :detective: